### PR TITLE
resume broker request for "get the app flow"

### PR DESCRIPTION
**Pull request by weijjia**
issue #485

### CODE REVIEW

**Comment by weijjia** [comment]
@sdabbiru Can you also take a look at this?

File: src/src/com/microsoft/aad/adal/ApplicationReceiver.java [hunk 1]
```diff
@@ -54,35 +58,48 @@
...
+        Logger.v(TAG + methodName, "ApplicationReceiver starts to save the request in shared preference.");
         SharedPreferences prefs = ctx.getSharedPreferences(INSTALL_REQUEST_TRACK_FILE,
                 Activity.MODE_PRIVATE);
+        
         if (prefs != null) {
             HashMap<String, String> parameters = StringExtensions.getUrlParameters(url);
-            if (parameters != null && parameters.containsKey(INSTALL_UPN_KEY)) {
+            if (parameters != null && parameters.containsKey(INSTALL_UPN_KEY)) 
+            {
+                Logger.v(TAG + methodName, "Coming redirect contains the UNP, setting it on the request for both loginhint and broker account name.");
```
Comment (kpanwar) [inline src/src/com/microsoft/aad/adal/ApplicationReceiver.java]: rename UNP to UPN.

File: src/src/com/microsoft/aad/adal/ApplicationReceiver.java [hunk 2]
```diff
@@ -131,45 +152,74 @@ public static void clearUserName(Context ctx) {
...
+        Logger.v(TAG + methodName, "Start resuming request in broker");
         Gson gson = new Gson();
-        AuthenticationRequest pendingRequest = gson.fromJson(request, AuthenticationRequest.class);
-        Intent intent = new Intent();
-        intent.setAction(Intent.ACTION_PICK);
-        intent.putExtra(AuthenticationConstants.Broker.BROKER_REQUEST, pendingRequest);
-        intent.putExtra(AuthenticationConstants.Broker.CALLER_INFO_PACKAGE, ctx.getPackageName());
-        intent.putExtra(AuthenticationConstants.Broker.BROKER_REQUEST_RESUME,
+        final AuthenticationRequest pendingRequest = gson.fromJson(request, AuthenticationRequest.class);
+        ExecutorService sThreadExecutor = Executors.newSingleThreadExecutor();
```
Comment (kpanwar) [inline src/src/com/microsoft/aad/adal/ApplicationReceiver.java]: Consider consolidating the executor service to a single instance and run a cachedThreadPool in later fixes. you an resolve this comment by filing an issue

File: src/AndroidManifest.xml [hunk 3]
```diff
@@ -12,6 +12,17 @@
         android:allowBackup="true">
         <activity android:name="com.microsoft.aad.adal.AuthenticationActivity" >
         </activity>
+        
+        <receiver
```
Comment (sdabbiru) [inline src/AndroidManifest.xml]: I suppose we need to update the ADAL developer documentation too, correct?

File: src/src/com/microsoft/aad/adal/ApplicationReceiver.java [hunk 4]
```diff
@@ -54,35 +58,48 @@
...
-                        .equalsIgnoreCase(
-                                "package:" + AuthenticationConstants.Broker.AZURE_AUTHENTICATOR_APP_PACKAGE_NAME)) {
-                    Logger.v(TAG, "Message is related to the broker");
+        final String methodName = ":onReceive";
+        if (intent.getAction().equals(Intent.ACTION_PACKAGE_ADDED)) 
+        {
+            Logger.v(TAG + methodName, "Application install message is received");
+            if (intent != null && intent.getData() != null) 
+            {
+                Logger.v(TAG + methodName, "Installing:" + intent.getData().toString());
```
Comment (sdabbiru) [inline src/src/com/microsoft/aad/adal/ApplicationReceiver.java]: Please update the log message... The installation already occurred.

File: src/src/com/microsoft/aad/adal/ApplicationReceiver.java [hunk 5]
```diff
@@ -54,35 +58,48 @@
...
+                final String receivedInstalledPackageName = intent.getData().toString();
+                if (receivedInstalledPackageName.equalsIgnoreCase("package:" + AuthenticationConstants.Broker.AZURE_AUTHENTICATOR_APP_PACKAGE_NAME) ||
+                        receivedInstalledPackageName.equalsIgnoreCase("package:" + AuthenticationSettings.INSTANCE.getBrokerPackageName()))
```
Comment (sdabbiru) [inline src/src/com/microsoft/aad/adal/ApplicationReceiver.java]: At some point, we need to move to the way we reference the azure_authenticator_app_package_name for CP too.

File: src/src/com/microsoft/aad/adal/AuthenticationContext.java [hunk 6]
```diff
@@ -1915,6 +1932,70 @@ public boolean isConnectionAvailable() {
...
+                else
+                {
+                    Logger.v(TAG + methodName, "No access token returned from broker resume request.");
+                    waitingRequest.mDelagete.onError(new AuthenticatorException("No access token returned from broker."));
+                }
+            }
+        }
+    }
+    
+    private AuthenticationResult processTokenResultFromBroker (final Intent intent)
```
Comment (sdabbiru) [inline src/src/com/microsoft/aad/adal/AuthenticationContext.java]: Where is this used?

File: src/src/com/microsoft/aad/adal/AuthenticationContext.java [hunk 7]
```diff
@@ -1915,6 +1932,70 @@ public boolean isConnectionAvailable() {
...
+            {
+                final String errorMessage = intent.getStringExtra(AuthenticationConstants.Browser.RESPONSE_ERROR_MESSAGE);
+                final String returnedErrorMessage = "ErrorCode: " + errorCode + " ErrorMessage" + errorMessage + getCorrelationInfoFromWaitingRequest(waitingRequest);
+                Logger.v(TAG + methodName, returnedErrorMessage);
+                waitingRequestOnError(waitingRequest, receivedWaitingRequestId, new AuthenticationException(ADALError.AUTH_FAILED, returnedErrorMessage));
+            }
+            else
+            {
+                Logger.v(TAG + methodName, "Received token result from broker reqest resume.");
+                final AuthenticationResult brokerResult = processTokenResultFromBroker(intent);
```
Comment (sdabbiru) [inline src/src/com/microsoft/aad/adal/AuthenticationContext.java]: Just so we do not forget - how do we ensure that only the intended app can process the token result and not some malicious app.

File: src/src/com/microsoft/aad/adal/ApplicationReceiver.java [hunk 8]
```diff
@@ -48,41 +54,60 @@
...
+                {
+                    Logger.v(TAG + methodName, receivedInstalledPackageName + " is installed, start sending request to broker.");
+                    
+                    installedPackageName = receivedInstalledPackageName.equalsIgnoreCase("package:" + AuthenticationConstants.Broker.AZURE_AUTHENTICATOR_APP_PACKAGE_NAME)? 
+                    		AuthenticationConstants.Broker.AZURE_AUTHENTICATOR_APP_PACKAGE_NAME : AuthenticationConstants.Broker.PACKAGE_NAME;
+                    
                     String request = getInstallRequestInthisApp(context);
-                    if (!StringExtensions.IsNullOrBlank(request)) {
-                        Logger.v(TAG, "Resume request in broker");
+                    if (!StringExtensions.IsNullOrBlank(request))
```
Comment (sdabbiru) [inline src/src/com/microsoft/aad/adal/ApplicationReceiver.java]: Shouldn't we check whether skipbroker is set to false before proceeding?

File: src/src/com/microsoft/aad/adal/ApplicationReceiver.java [hunk 9]
```diff
@@ -131,45 +160,102 @@ public static void clearUserName(Context ctx) {
...
-            ctx.startActivity(intent);
+                if (isIntentSafe) 
+                {
+                    if (!isRunningInForeground(context, installedPackageName))
+                    {
+                        Logger.v(TAG + methodName, "Broker " + installedPackageName + " is not running in foreground yet. "
+                            + "Thread will be sleeping for 2 seconds to wait for broker package going foreground.");
+                        try 
+                        {
+                            Thread.sleep(2000);
```
Comment (sdabbiru) [inline src/src/com/microsoft/aad/adal/ApplicationReceiver.java]: I have a older, slower device. We should test this on it. In any case, I think we should do retries with shorter intervals.

File: src/src/com/microsoft/aad/adal/AuthenticationContext.java [hunk 10]
```diff
@@ -1915,6 +1934,72 @@ public boolean isConnectionAvailable() {
...
+                    
+                    if (StringExtensions.IsNullOrBlank(userId))
+                    {
+                        userId = authenticationRequest.getLoginHint();
+                    }
+
+                    try
+                    {
+                        final AuthenticationResult authResult = acquireTokenSilentSync(authenticationRequest.getResource(), authenticationRequest.getClientId(), userId);
+                        waitingRequest.mDelagete.onSuccess(authResult);
```
Comment (sdabbiru) [inline src/src/com/microsoft/aad/adal/AuthenticationContext.java]: Please add some logging here...

File: src/src/com/microsoft/aad/adal/AuthenticationContext.java [hunk 11]
```diff
@@ -1915,6 +1934,72 @@ public boolean isConnectionAvailable() {
...
+                    }
+                    catch (Exception exception)
+                    {
+                        waitingRequestOnError(waitingRequest, receivedWaitingRequestId, new AuthenticationException(ADALError.AUTH_FAILED, exception.getMessage()));
```
Comment (sdabbiru) [inline src/src/com/microsoft/aad/adal/AuthenticationContext.java]: logging...

File: src/src/com/microsoft/aad/adal/ApplicationReceiver.java (hunk 8 above)
Comment (weijjia) [inline src/src/com/microsoft/aad/adal/ApplicationReceiver.java]: Right, we should check if we can switch to broker.

File: src/src/com/microsoft/aad/adal/ApplicationReceiver.java (hunk 2 above)
Comment (weijjia) [inline src/src/com/microsoft/aad/adal/ApplicationReceiver.java]: https://github.com/AzureAD/azure-activedirectory-library-for-android/issues/507

File: src/src/com/microsoft/aad/adal/ApplicationReceiver.java (hunk 9 above)
Comment (weijjia) [inline src/src/com/microsoft/aad/adal/ApplicationReceiver.java]: sure. I will give it a try tomorrow.

File: src/AndroidManifest.xml (hunk 3 above)
Comment (weijjia) [inline src/AndroidManifest.xml]: yup, i will add it in the readme when we're ready to ship the new broker. I also want to know the version of the two broker apps with the PRT feature and have a note on the readme in order not to confuse people.

**Comment by weijjia** [comment]
@sdabbiru I want to hold on merging this PR. I prefer to have the exception handling PR merged and released for OneDrive, have FoCI PR merged  and released. And then we are in the good state for merging this one if there are no other outstanding comments on this one.

File: src/src/com/microsoft/aad/adal/ApplicationReceiver.java [hunk 12]
```diff
@@ -131,45 +163,101 @@ public static void clearUserName(Context ctx) {
...
-        AuthenticationRequest pendingRequest = gson.fromJson(request, AuthenticationRequest.class);
-        Intent intent = new Intent();
-        intent.setAction(Intent.ACTION_PICK);
-        intent.putExtra(AuthenticationConstants.Broker.BROKER_REQUEST, pendingRequest);
-        intent.putExtra(AuthenticationConstants.Broker.CALLER_INFO_PACKAGE, ctx.getPackageName());
-        intent.putExtra(AuthenticationConstants.Broker.BROKER_REQUEST_RESUME,
+        final AuthenticationRequest pendingRequest = gson.fromJson(request, AuthenticationRequest.class);
+        ExecutorService sThreadExecutor = Executors.newSingleThreadExecutor();
+        
+        sThreadExecutor.submit(new Callable<Object>()
```
Comment (agamemnum) [inline src/src/com/microsoft/aad/adal/ApplicationReceiver.java]: Don't use `submit` - use `execute` here

File: src/src/com/microsoft/aad/adal/AuthenticationContext.java [hunk 13]
```diff
@@ -1916,6 +1935,87 @@ public boolean isConnectionAvailable() {
...
+                            waitingRequest.mDelagete.onSuccess(authResult);
+                        }
+                        else
+                        {
+                            Logger.v(TAG + methodName, "acquireTokenSilentSync completes successfully, but token is not returned.");
+                            waitingRequestOnError(waitingRequest, receivedWaitingRequestId, new AuthenticationException(ADALError.AUTH_FAILED, 
+                                    "acquireTokenSilentSync completes successfully, but token is not returned." + getCorrelationInfoFromWaitingRequest(waitingRequest)));
+                        }
+                    }
+                    catch (Exception exception)
```
Comment (agamemnum) [inline src/src/com/microsoft/aad/adal/AuthenticationContext.java]: Please no `catch Exception`

File: src/src/com/microsoft/aad/adal/AuthenticationContext.java [hunk 14]
```diff
@@ -823,6 +830,31 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
...
+                        
+                        @Override
+                        public void run() {
+                            if (!mBrokerResumeResultReceiver.isResultReceivedFromBroker()) {
+                                Logger.v(TAG + "onActivityResult", "BrokerResumeResultReceiver doesn't receive result from "
+                                        + "broker within 10 minuites, unregister the receiver and cancelling the request");
+                                
+                                (new ContextWrapper(mContext)).unregisterReceiver(mBrokerResumeResultReceiver);
+                                waitingRequestOnError(waitingRequest, requestId, new AuthenticationCancelError("Broker doesn't "
+                                        + "return back the result within 5 minuites"));
```
Comment (sdabbiru) [inline src/src/com/microsoft/aad/adal/AuthenticationContext.java]: Should be 10 mins or 5?

File: src/src/com/microsoft/aad/adal/BasicWebViewClient.java [hunk 15]
```diff
@@ -253,9 +255,15 @@ public void run() {
             ApplicationReceiver.saveRequest(mCallingContext, mRequest, url);
             HashMap<String, String> parameters = StringExtensions
                     .getUrlParameters(url);
+            prepareForBrokerResumeRequest();
+            // Wait for the calling activity to receive the result from prepareForBrokerResumeRequest
+            try {
+                Thread.sleep(1000);
```
Comment (sdabbiru) [inline src/src/com/microsoft/aad/adal/BasicWebViewClient.java]: Please put a comment as to why the 1 sec delay is necessary.

File: src/src/com/microsoft/aad/adal/BasicWebViewClient.java [hunk 16]
```diff
@@ -253,9 +255,15 @@ public void run() {
...
+            } catch (InterruptedException e) {
+                Logger.v(TAG + ":shouldOverrideUrlLoading", "Error occure when having thread sleeping for 1 second");
```
Comment (sdabbiru) [inline src/src/com/microsoft/aad/adal/BasicWebViewClient.java]: Please update the log.

File: src/src/com/microsoft/aad/adal/AuthenticationContext.java (hunk 14 above)
Comment (weijjia) [inline src/src/com/microsoft/aad/adal/AuthenticationContext.java]: should be 10. Will have it updated.

File: src/src/com/microsoft/aad/adal/BasicWebViewClient.java (hunk 16 above)
Comment (weijjia) [inline src/src/com/microsoft/aad/adal/BasicWebViewClient.java]: what do you need me to update to? error log or?

**Comment by sdabbiru** [comment]
:shipit:

**Pull request merged**